pkg/cli/image/mirror/file.go

```diff
@@ -0,0 +1,441 @@
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		return distribution.Descriptor{}, err
+	}
+	if fi.Mode()&os.ModeSymlink == 0 {
+		return distribution.Descriptor{}, fmt.Errorf("not a symlink")
+	}
+	target, err := os.Readlink(path)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+	return distribution.Descriptor{
+		Digest: godigest.Digest(filepath.Base(target)),
+		Size:   fi.Size(),
+	}, nil
+}
+
+// Tag associates the tag with the provided descriptor, updating the
+// current association, if needed.
+func (s *fileTagStore) Tag(ctx context.Context, tag string, desc distribution.Descriptor) error {
+	panic("not implemented")
+}
+
+// Untag removes the given tag association
+func (s *fileTagStore) Untag(ctx context.Context, tag string) error {
+	panic("not implemented")
+}
+
+// All returns the set of tags managed by this tag service
+func (s *fileTagStore) All(ctx context.Context) ([]string, error) {
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "manifests")
+	fis, err := ioutil.ReadDir(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	var names []string
+	for _, fi := range fis {
+		if fi.Mode()&os.ModeSymlink != 0 {
+			names = append(names, filepath.Base(fi.Name()))
+		}
+	}
+	return names, nil
+}
+
+// Lookup returns the set of tags referencing the given digest.
+func (s *fileTagStore) Lookup(ctx context.Context, digest distribution.Descriptor) ([]string, error) {
+	panic("not implemented")
+}
+
+type fileManifestService struct {
+	r *fileRepository
+}
+
+// Exists returns true if the manifest exists.
+func (s *fileManifestService) Exists(ctx context.Context, dgst godigest.Digest) (bool, error) {
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "manifests", dgst.String())
+	fi, err := os.Lstat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if fi.IsDir() || fi.Mode()&os.ModeSymlink != 0 {
+		return false, fmt.Errorf("not a file")
+	}
+	return true, nil
+}
+
+// Get retrieves the manifest specified by the given digest
+func (s *fileManifestService) Get(ctx context.Context, dgst godigest.Digest, options ...distribution.ManifestServiceOption) (distribution.Manifest, error) {
+	path := filepath.Join(s.r.basePath, "v2", s.r.repoPath, "manifests", dgst.String())
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var versioned manifest.Versioned
+	if err = json.Unmarshal(data, &versioned); err != nil {
+		return nil, err
+	}
+
+	manifest, desc, err := distribution.UnmarshalManifest(versioned.MediaType, data)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(5).Infof("Read manifest %T from %s: %v", manifest, path, desc)
+	return manifest, nil
+}
+
+func atomicWrite(path string, data []byte) error {
+	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(path+".download", data, 0600); err != nil {
```

Comment:
is it expected that the path can never be a directory path?

Reply:
yes, in that there is only one level of file leaves allowed (no dir in this structure that has files has subduers